### ISSUE TITLE
[solc] ASTUtils.cpp throws ICE "Identifier is not referencing a variable declaration"

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#8418]
## Description

```
contract C {
  bool constant d = this;
  function f() {
    assembly { y := d }
  }
}
```

throws

https://github.com/ethereum/solidity/blob/a3bc752e5326b6c7341bba545044fd116e097942/libsolidity/ast/ASTUtils.cpp#L33

## Environment

- Compiler version: latest develop
- Target EVM version (as per compiler settings): NA

## Steps to Reproduce

```
$ solc <input.sol>
Internal compiler error during compilation:
/home/bhargava/work/github/solidity/libsolidity/ast/ASTUtils.cpp(33): Throw in function const solidity::frontend::VariableDeclaration *solidity::frontend::rootVariableDeclaration(const solidity::frontend::VariableDeclaration &)
Dynamic exception type: boost::exception_detail::clone_impl<solidity::langutil::InternalCompilerError>
std::exception::what: Identifier is not referencing a variable declaration
[solidity::util::tag_comment*] = Identifier is not referencing a variable declaration
```

**Comment by chriseth** [comment]
@a3d4 you might want to take a look here.

**Comment by a3d4** [comment]
```c++
VariableDeclaration const* rootVariableDeclaration(VariableDeclaration const& _varDecl)
{
	solAssert(_varDecl.isConstant(), "Constant variable expected");

	VariableDeclaration const* rootDecl = &_varDecl;
	Identifier const* identifier;
	while ((identifier = dynamic_cast<Identifier const*>(rootDecl->value().get())))
	{
		auto referencedVarDecl = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration);
		solAssert(referencedVarDecl && referencedVarDecl->isConstant(), "Identifier is not referencing a variable declaration");
		rootDecl = referencedVarDecl;
	}
	return rootDecl;
}
```

The issue is that `VariableDeclaration` is expected, while `this` is a `MagicVariableDeclaration`. Similarly, `referencedVarDecl->isConstant()` condition of the `assert` fails for

```
contract C {
  bool c = false;
  bool constant d = c;
  function f() {
    assembly { y := d }
  }
}
```

Either we should not touch "already broken constants" during assembly parsing, or we should replace the `assert` with an `if` or alike. Otherwise we are asserting on invalid contract code.

`rootVariableDeclaration()` is called twice (`TypeChecker::visit(InlineAssembly)`, `ContractCompiler::visit(InlineAssembly)`).

Instead of `assert`, `rootVariableDeclaration` can return `nullptr`. Checking for that `nullptr` in callers seem to compensate the removed `assert`s.

See PR #8480.

**Comment by chriseth** [comment]
Fixed by #8480
